pink/solve_ik.py

```diff
@@ -37,29 +41,40 @@ def __compute_qp_objective(
             `[cost]^2 / [tangent]` where `[tangent]` is "the" unit of robot
             velocities. Improves numerical stability, but larger values slow
             down all tasks.
+        dt: Integration timestep in [s].
+        barriers: Collection of barriers.
 
     Returns:
         Pair :math:`(H, c)` of Hessian matrix and linear vector of the QP
         objective.
     """
+    barriers = barriers if barriers is not None else []
     H = damping * configuration.tangent.eye
     c = configuration.tangent.zeros.copy()
     for task in tasks:
         H_task, c_task = task.compute_qp_objective(configuration)
         H += H_task
         c += c_task
+
+    for barrier in barriers:
+        H_cbf, c_cbf = barrier.compute_qp_objective(configuration)
```

Comment:
As far as I understand, barrier objectives so far (also the spherical and configuration barriers from your fork) have a zero ``safe_policy`` vector, so they are otherwise unrelated to the barrier itself.

If that's the case, I'd be in favor of refactoring one step further:

- Create a separate task for the ``safe_policy`` with variable $\gamma(q)$
- Remove ``compute_qp_objective`` from barriers

That way, we would have a clean separation between "tasks = QP objective" and "barriers = QP constraints".

What do you think?

Reply:
In my opinion, the notion of barrier is quite general, and someone might find more effective safe policy rather than 0. And if this would be the case, then he might need to use barrier value and/or its jacobian. Therefore, it still makes sense to me to define both constraint and gain in one class.

Moreover, in my opinion it is perfectly fine that some components of the problem define both objective and constraints. As an example, I should mention Control Lyapunov Functions, which are usually present in quadratic programs as follows (sorry for image, latex refused to parse it for some reason):
<img width="373" alt="Screenshot 2024-05-18 at 18 35 15" src="https://github.com/stephane-caron/pink/assets/28687492/47678483-5913-4aea-9eaa-ddd61f2be13d">

Anyhow, I am not sure what is the best way to handle it, and if you persist that they should be separate, I am ready to consider possible solutions for that. Also I'd like to hear from @simeon-ned his opinion on this matter.

Reply:
@stephane-caron, as I understand it, this is just another form of Tikhonov-like regularization that you have already introduced in the `solve_ik` or `lm_damping` in tasks, but now this is configuration-dependent. So, in my view, there are two options:

1. Leave it as it is (but perhaps rename it to `compute_safe_displacement`/`safe_velocity`).
2. Remove this from barriers and add a separate `damping` task, and specifically discuss in the documentation that the user needs to introduce this damping to prevent chattering.

Personally, I agree with @domrachev03. It seems to me that `safe_displacement` is tightly coupled with its own barrier and its Jacobian, so it is more convenient to define this within the barrier framework rather than as a separate task. It would be somewhat confusing to force the user to always add an additional `safe_displacement` task for each barrier.

However, if you think that separating them would be clearer and better suit the style of PINK, we are open to considering this approach.

Please let me know if you have any questions or further thoughts on this matter.

Reply:
Thank you both for your feedback. I see how in the long run barriers can factor computations, with e.g. the Jacobian implemented once and used in both cost and constraint. I agree with that :+1: 

Now I wonder if the base-class implementation of `compute_qp_objective` (with an underlying `compute_self_displacement`) will be generic enough for all barriers, or if the base class should just do nothing and each barrier should be responsible for implementing its own objective?

That's not a blocker, we can move forward with the current implementation. (We can always revise the decision later on when there are more barriers to look at.)

Reply:
This method is overridable anyway, so if some implementation of barriers really needs to implement something special, no one stops them.

Yet, all barriers we've considered and implemented did not require any special treatment, and worked pretty well.